File: src/leetcode/P122_BestTimetoBuyandSellStockII.java
package leetcode;

/**
    Say you have an array for which the ith element is the price of
    a given stock on day i.

	Design an algorithm to find the maximum profit. 
	You may complete as many transactions as you like 
	(ie, buy one and sell one share of the stock multiple times). 
	However, you may not engage in multiple transactions at the same time 
	(ie, you must sell the stock before you buy again).
 */

/**
 * @author      zxwtry
 * @email       zxwtry@qq.com
 * @project     OJ
 * @package     leetcode
 * @file        P122_BestTimetoBuyandSellStockII.java
 * @type        P122_BestTimetoBuyandSellStockII
 * @date        2017年5月8日 下午2:14:39
 * @details     AC 2ms 10.44%
 */
public class P122_BestTimetoBuyandSellStockII {
	static class Solution {
	    public int maxProfit(int[] prices) {
	    	if (prices == null) {
	    		return 0;
	    	}
	        int ans = 0;
	        for (int i = prices.length-1; i > 0; i --) 
	            ans += Math.max(0, prices[i] - prices[i-1]);
	        return ans;
	    }
	}
}

File: python/P122_BestTimeToBuyAndSellStockII.py
#coding=utf-8

'''
    url: leetcode.com/problems/best-time-to-buy-and-sell-stock-ii
    @author:     zxwtry
    @email:      zxwtry@qq.com
    @date:       2017年5月5日
    @details:    Solution:  56ms 39.52%
'''

class Solution(object):
    def maxProfit(self, p):
        """
        :type p: List[int]
        :rtype: int
        """
        return  sum(max((p[i]-p[i-1]), 0) for i in range(1, len(p)))

if __name__ == "__main__":
    p = [1, 2, 3, 6, 4, 7]        
    print(Solution().maxProfit(p))